### PR TITLE
Propagate dependencies in CMake configuration file

### PR DESCRIPTION
**Pull request by juanjosegarciaripoll**
## Pull request purpose

The CMake files that are installed by arpack are non-standard and do not enumerate the library dependencies. This is critical where arpack is not built as a shared library (e.g. Windows), but more generally it is not considered good practice. This way, users can just use
```
target_add_libraries(myproject PRIVATE arpackng)
```
instead of having themselves to build something like
```
find_package(BLAS REQUIRED)
[... code for older versions of cmake's findBlas ...]
find_package(LAPACK REQUIRED)
[... code for older versions of cmake's findBlas ...]
target_add_libraries(myproject PRIVATE arpack BLAS::BLAS LAPACK::LAPACK)
```

## Detailed changes proposed in this pull request

- Rewrite arpack-ng-config.cmake to find its dependencies using standard tools (find_dependency) and rely on a targets file
- Create this targets file as arpackngTargets.cmake, relying on CMake to create proper targets, with paths to include and library directories
- Declare BLAS::BLAS and LAPACK::LAPACK and MPI::Fortran as public dependencies of arpack and parpack
- Define ARPACK::ARPACK and PARPACK::PARPACK as aliases

### CODE REVIEW

**Comment by juanjosegarciaripoll** [comment]
This patch leaves some unresolved issues:
- This fix is compatible with the old CMake configuration, which adds to include/arpack-ng${suffixes} to the include path. However, this is not standard practice, polluting the C/C++ namespace with include files that should rather be used as in `#include <arpack-ng/...>`.
- The tests in cmake/ do not verify that the CMake file is found with find_package() using standard locations
- The CMake file uses nonstandard binary file relocation techniques for all tests and examples. Those techniques do not work with Windows Mingw64, leaving the examples broken.
- Instead, CMakeLists.txt files should be created in the subdirectories where examples and tests live, as it is done with autotools.
- There is an abuse of global variables in the CMake file that can be corrected by defining common repetitive tasks and global identification of libraries as separate functions, as it is customary in modern CMake practices.

**Comment by fghoussen** [comment]
Could you update the `CHANGES` file? No time to have a close look but sounds good! Will try to have a look this week-end.

> The tests in cmake/ do not verify that the CMake file is found with find_package() using standard locations

I know, I designed it! :) This was the best I could get at the time trying to test the way the package is installed (trying to test use of  `find_package` - we had lots of problems around this and no way to test so blind pushing was the rule). Would be good to be improved by future PR

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: Does it means now, using find_dependencies, all stuffs related to blas / lapack / mpi could be suppressed? If relevant, could you impact the test `cmake/tstCMakeInstall.sh.in ` in this PR?

File: CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,12 @@ set(SYMBOLSUFFIX ""
     CACHE STRING "suffix to add to ARPACK, BLAS and LAPACK function names")
 option(INTERFACE64 "use the 64-bit integer interface (ILP64) for ARPACK, BLAS and LAPACK")
 
+# 'make install' to the correct location, and also define
+# paths for target_include_directories and target_link_libraries
+include(GNUInstallDirs)
+set(ARPACK_INSTALL_INCLUDEDIR "${CMAKE_INSTALL_INCLUDEDIR}/arpack-ng${ITF64SUFFIX}")
+set(ARPACK_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/arpackng${LIBSUFFIX}${ITF64SUFFIX}")
```
Comment (fghoussen) [inline CMakeLists.txt:48]: Set `set(ARPACK_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/arpack-ng${LIBSUFFIX}${ITF64SUFFIX}")` _with_ dash for consistency? @sylvestre: what do you think?

File: CMakeLists.txt [hunk 2]
```diff
@@ -377,18 +383,37 @@ endif()
 # use -DBUILD_SHARED_LIBS=ON|OFF to control static/shared
 add_library(arpack ${arpackutil_STAT_SRCS} ${arpacksrc_STAT_SRCS} ${arpacksrc_ICB})
 
-target_link_libraries(arpack BLAS::BLAS)
-target_link_libraries(arpack LAPACK::LAPACK ${EXTRA_LDFLAGS})
+target_link_libraries(arpack
+  PUBLIC
+  $<INSTALL_INTERFACE:$<IF:$<BOOL:${BUILD_SHARED_LIBS}>,,BLAS::BLAS>>
+  $<INSTALL_INTERFACE:$<IF:$<BOOL:${BUILD_SHARED_LIBS}>,,LAPACK::LAPACK>>
+  $<BUILD_INTERFACE:BLAS::BLAS>
```
Comment (fghoussen) [inline CMakeLists.txt:390]: @juanjosegarciaripoll: could you just explain what you do here? difference between  INSTALL_INTERFACE and BUILD_INTERFACE? (I try to improve my knowldge of cmake ! :))

File: CMakeLists.txt [hunk 3]
```diff
@@ -788,6 +820,7 @@ configure_file(pkg-config/arpackSolver.pc.in "${PROJECT_BINARY_DIR}/arpackSolver
 
 
 install(TARGETS arpack
+    EXPORT arpackngTargets
```
Comment (fghoussen) [inline CMakeLists.txt:823]: @juanjosegarciaripoll: could you just explain what you do here? not used to target export: why is this needed? (I try to improve my knowldge of cmake ! :))

File: cmake/arpackng-config.cmake.in [hunk 4]
```diff
@@ -12,17 +12,29 @@
...
 
-# Create parpack targets.
-set(MPI "@MPI@")
-if (MPI)
-    add_library(PARPACK::PARPACK INTERFACE IMPORTED)
-    set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_INCLUDEDIR@/arpack-ng@ITF64SUFFIX@")
-    set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_LINK_DIRECTORIES    "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBDIR@")
-    set_target_properties(PARPACK::PARPACK PROPERTIES INTERFACE_LINK_LIBRARIES      "parpack@LIBSUFFIX@@ITF64SUFFIX@")
-endif ()
+include("${CMAKE_CURRENT_LIST_DIR}/arpackngTargets.cmake")
```
Comment (fghoussen) [inline cmake/arpackng-config.cmake.in:35]: @juanjosegarciaripoll: in connection with previous comment, just to understand, a .cmake file is created here, right? what does it contain and why? How is it used by cmake? (I try to improve my knowldge of cmake ! :))

File: CMakeLists.txt (hunk 1 above)
Comment (juanjosegarciaripoll) [reply CMakeLists.txt:48]: I just copied the original configuration. Making this change should lead to incompatibilities, specially if it is also not done consistently with the autoconf scripts. I leave it to your judgement to decide.

File: CMakeLists.txt (hunk 2 above)
Comment (juanjosegarciaripoll) [reply CMakeLists.txt:390]: BUILD_INTERFACE macros are replaced when the library and its components (e.g. tests) are compiled. In this case, we are declaring that the library depends on BLAS::BLAS, LAPACK::LAPACK and that the components that are built with it (the tests) will get this dependency automatically transferred to them (PUBLIC).

INSTALL_INTERFACE is a macro that is only expanded when the library is finally exported, in this case to an arpacksTarget.cmake file. Through this text we make the library export its dependency on BLAS and LAPACK only when it is built as a statically linked library. I believe this is the minimum that must be done.

File: CMakeLists.txt (hunk 3 above)
Comment (juanjosegarciaripoll) [reply CMakeLists.txt:823]: We instruct CMake to collect all the declarations we have made above (what include directories arpack depends on, what libraries it depends on, what link flags it needs to be built, etc) and create a file that exports that information about this particular target. A similar statement below applies to parpack as well. This information is the robust, relocatable way to export include and link flags from CMake projects. The developer does not need to write variables as it was done before.

File: cmake/arpackng-config.cmake.in (hunk 4 above)
Comment (juanjosegarciaripoll) [reply cmake/arpackng-config.cmake.in:35]: As mentioned above, this file describes the two targets that this project create, 'arpack' and 'parpack', creating them as CMake objects with the right values for all properties (include directories, link libraries, link flags, etc). It is done so in a way that the target is immediately available once this file is included from arpackng-config.cmake. It is the modern way of exporting targets.

**Comment by juanjosegarciaripoll** [comment]
> @juanjosegarciaripoll: Does it means now, using find_dependencies, all stuffs related to blas / lapack / mpi could be suppressed? If relevant, could you impact the test `cmake/tstCMakeInstall.sh.in ` in this PR?

It depends. I am unfamiliar with the fortran files that are linked in the test. If those fotran files use BLAS or LAPACK functions themselves, they need to declare their own dependency on those libraries in the files that `tstCMakeInstall.sh.in` creates.

When Arpack is built as a statically linked library, we definitely must export those dependencies publicly, so that applications linked with arpack also link with the right libraries. A side effect is that the user's objects have immediate access to BLAS and LAPACK, but that is not always the case -- for instance, when Arpack is built as a shared library, the dependency on BLAS and LAPACK is not transitively passed to the software that uses arpack.

A rule of thumb is that when a target (in this case the fortran tests) calls functions from a given library it must declare it as a PUBLIC or PRIVATE dependency.

**Comment by juanjosegarciaripoll** [comment]
I don't know what happened, but one of the comments to this pull request turned into a commit where tstCMakeInstall.sh was edited. I have not opened permissions to do so, but I have reverted the change, because as I said before, if `dnband.f` uses BLAS, it must declare its own dependency, cannot be propagated from `arpackng` (arpack only needs to propagate the dependencies it needs to be linked itself).

File: CMakeLists.txt (hunk 1 above)
Comment (fghoussen) [reply CMakeLists.txt:48]: > I just copied the original configuration.

Yes, got it.

> I leave it to your judgement to decide.

Weird ?!... From CI on master https://github.com/opencollab/arpack-ng/actions/runs/2895871392/jobs/4606355121#step:6:289, cmake seems to install include and libs in an arpack-ng (_with_ dash) directory so we should have `set(ARPACK_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/arpack-ng${LIBSUFFIX}${ITF64SUFFIX}")`, where autotools produces this install https://github.com/opencollab/arpack-ng/actions/runs/2895871392/jobs/4606354821#step:6:1430

Anyway, I'll set `set(ARPACK_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/arpack-ng${LIBSUFFIX}${ITF64SUFFIX}")`, _with_ dash, for consistency if the install tests (`tstCMakeInstall.sh` and tstAutotoolsInstall.sh``) pass: this way at least what worked before still work. I'll push this modification in a commit on top of yours: you can retrieve it by pulling at your side, and then, you'll can push on top of me.

File: CMakeLists.txt (hunk 2 above)
Comment (fghoussen) [reply CMakeLists.txt:390]: Thanks for clarifications!

File: CMakeLists.txt (hunk 3 above)
Comment (fghoussen) [reply CMakeLists.txt:823]: Thanks for clarifications!

File: cmake/arpackng-config.cmake.in (hunk 4 above)
Comment (fghoussen) [reply cmake/arpackng-config.cmake.in:35]: Thanks for clarifications!

**Comment by fghoussen** [comment]
Can you update the changelog?

> It depends. I am unfamiliar with the fortran files that are linked in the test.

I've just checked: the fortran file need BLAS, LAPACK so the commit I pushed (to test) on top of you needed to be reverted. 

> don't know what happened, but one of the comments to this pull request turned into a commit where tstCMakeInstall.sh was edited.

I pushed on top of you but you pushed-force after me so all is OK now!

**Comment by fghoussen** [comment]
Sounds like [b1fa008](https://github.com/opencollab/arpack-ng/pull/377/commits/b1fa0089696fe2d069bfc44c4eab9f844e1a30f2) breaks `ubuntu_latest_cmake_install` so it needs to be reverted. Not sure to get why anyway...

**Comment by fghoussen** [comment]
Cmake needs all file in `cmake/arpackng` dir
image
and breaks if we have 
image
So [b1fa008](https://github.com/opencollab/arpack-ng/pull/377/commits/b1fa0089696fe2d069bfc44c4eab9f844e1a30f2) needs to be reverted

File: CMakeLists.txt (hunk 1 above)
Comment (juanjosegarciaripoll) [reply CMakeLists.txt:48]: No, PLEASE, if you do so, find_package(arpackng) will NOT find the CMake configuration files! We have discussed this already. `find_package(arpackng)` searchers for `/usr/lib/cmake/arpackng*` directories. If you change the installation to `/usr/lib/cmake/arpack-ng`, then CMake will not find the `arpackng-config.cmake` file

Comment (juanjosegarciaripoll) [reply CMakeLists.txt:48]: As I mentioned in the other pull request, this is one of the critical fixes in this patch. If this is changed, then the user must supply CMake with the installation directory for Arpack, which is *not* how things are done when using CMake. Users should be able to simply use `find_package(arpackng)` and let the build system find the packages. This is critical for allowing `arpackng` being built in systems such as `vcpkg` which is what I am trying to do.

Comment (juanjosegarciaripoll) [reply CMakeLists.txt:48]: Changing the installation directory to `arpackng` just for the CMake files does not lead to any inconsistency, because the former CMake installation was simply broken.

Comment (juanjosegarciaripoll) [reply CMakeLists.txt:48]: It must also be remarked that the autoconf tools are not capable of building useful CMake files at this moment -- they will never be able to reproduce what CMake is doing with their targets. Moreover, they are never installed in any of the platforms where I have seen arpack distributed; only the pkg-config file is available (e.g Debian, Ubuntu), so i do not think there is a need to keep both of them in sync in this aspect.

**Comment by fghoussen** [comment]
Can you update the changelog? And I merge this by the week-end.
Changelog helps to track important changes like this one is.

> No, PLEASE, if you do so, find_package(arpackng) will NOT find the CMake configuration files!... This is critical for allowing arpackng being built in systems such as vcpk

OK, got it now! :) Lots of stuff to get right

> autoconf tools are not capable of building useful CMake files

They don't.

**Comment by juanjosegarciaripoll** [comment]
I uploaded a change to CHANGES detailing what has been done.

Now, with these changes, arpack-ng builds and runs under vcpkg. I am very happy that the port requires no hack or diff. It is temporarily available here https://github.com/juanjosegarciaripoll/vcpkg-ports and can be built using
```
vcpkg install arpackng --overlay-ports=path-to-vcpkg-ports/ports/arpackng
```
However, as soon as it is committed, I will be able to do a pull request to vcpkg to include this recipe.

This said, I am experiencing somewhat troubling instabilities and lack of convergence with arpack that did not happen with the earlier version I was stuck with. It is a bit troublesome. I have opened an issue, but it is very easy to test.

**Comment by fghoussen** [comment]
@juanjosegarciaripoll: thanks!
